src/model/Request.php

```diff
@@ -5,18 +5,45 @@
 // hnusna skareda staticka vec! fuj!
 class Request
 {
-    public static function getIp()
+    public static function getIp(): string
     {
-        if (!empty($_SERVER['HTTP_CLIENT_IP'])) {
-            $ip = $_SERVER['HTTP_CLIENT_IP'];
-        } elseif (!empty($_SERVER['HTTP_X_FORWARDED_FOR']) && $_SERVER['HTTP_X_FORWARDED_FOR'] != '127.0.0.1') {
-            $ip = $_SERVER['HTTP_X_FORWARDED_FOR'];
-        } elseif (!empty($_SERVER['REMOTE_ADDR'])) {
-            $ip = $_SERVER['REMOTE_ADDR'];
-        } else {
-            $ip = 'cli';
+        $serverKeysToCheck = [
+            'HTTP_R_FORWARDED_FOR', // special header for custom IP if needed
```

Comment:
This custom header is a little bit worrying (it looks too custom). Wouldn't you be able to use any of other headers to set custom IP if needed?

Reply:
I could, but I can't really ensure, that no one (some proxy or load balancer for example) on the way will change it.. This way we can be sure, that the value of this header is relevant..

Reply:
This would probably open window for vulnerabilities.. Anybody could fake an IP address with this (without need of using VPN or some other service). I'm definitely not an expert on this, but there's probably a reason why it's always a bit complicated.

We've checked Symfony implementation - https://github.com/symfony/http-foundation/blob/master/Request.php#L793 -  and also found out that Nette has something similar (including trusted proxies) - https://github.com/nette/http/blob/master/src/Http/RequestFactory.php#L328.

How about we try the Nette original implementation?

Reply:
That is not easy to read code :) Using the nette original implementation means, that the Request class will have to be changed to a servis right? I think, faking the IP address is not really a problem with the current implementation, if someone requests the www server, he can override the values of these headers. Since there is no standard, it is hard to tell if it will be or how it will be changed on the way...

Reply:
Yes, it would have to be a service. To be more precise, you'd use Nette's `Request` directly (and inject it via DI) and ignore our helper which shouldn't probably even exist.

Regarding safety, even our current implementation is probably not completely OK. That's why I'd probably dump it in favor of Nette's `Request`. What's missing in our implementation are trusted proxies that could solve spoofing and wouldn't use "any" IP address that comes in the headers.

What I'm really trying to avoid is to add completely custom `HTTP_R_FORWARDED_FOR` header. I'm definitely not familiar with these kind of server settings, but here I'd more trust Nette/Symfony implementation than ours.

Even Rails are only checking `REMOTE_ADDR`, `CLIENT_IP` and `X_FORWARDED_FOR`: https://github.com/rails/rails/blob/master/actionpack/lib/action_dispatch/middleware/remote_ip.rb#L112  
(see also the comments in the files and linked URLs; these headers are not RFC standard, but they're widely agreed on across the most popular web servers)

Reply:
The only specific method is `isApi` in this class, but it is not used anywhere I think (I found it only in the AccessToken class in the users module, but it is commented out).. The main reason behind the custom header was, that there is a scenario, when I want to "spoof" the IP address. Like the mobile app, which I described. When the mobile app does not communicate with remp directly, then the "middle man" (a so called proxy, which communicates with remp api on behalf of the mobile app) needs to forward the users IP address to remp somehow, otherwise all app users will have the proxy IP. I can try setting/adding the user IP as client IP or x_forwarded_for, hopefully it will work and no one will change it on the way. So you say, it will be better to completely eliminate the usage of this class, and use nette request in the whole users module? Or maybe a wrapper around it with some extra helper functions which could be useful elsewhere?

Reply:
On proxy for mobile app application: Setting `X_FORWARDED_FOR` on the side of proxy seems as a much cleaner solution to me than using custom header that would work only in this application. As `X_FORWARDED_FOR` is an agreed way to do it, there's a good chance you'll be able to reuse it elsewhere if necessary.

On whether to use this class or Nette's request: If Nette's request does what you need it to do, it's better to use it directly. I can't really imagine a scenario where we would need to wrap it. The safest way to handle this is to ask DI for  `Nette\Http\IRequest` and let DI to provide you the implementation. In case we need to wrap it in the future, we'll do it in a way that your implementation wouldn't break. We'll probably be changing this to Nette's request IP address resolving as well.